packages/runtime/src/runtime/hooks/useEnsembleAction.tsx

```diff
@@ -120,6 +131,174 @@ export const useInvokeApi: EnsembleActionHook<InvokeAPIAction> = (action) => {
+    input.accept =
+      action?.allowedExtensions?.map((ext) => ".".concat(ext))?.toString() ||
+      "*/*";
+
+    input.onchange = (event: Event): void => {
+      const selectedFiles =
+        (event.target as HTMLInputElement).files || undefined;
+
+      if (selectedFiles) {
+        setFiles(selectedFiles);
+        onCompleteAction?.callback();
+      }
+    };
+
+    return input;
+  }, [action?.allowMultiple, action?.allowedExtensions, onCompleteAction]);
+
+  useEffect(() => {
+    return () => {
+      inputEl.remove();
+    };
+  }, [inputEl]);
+
+  const callback = useCallback((): void => {
+    try {
+      inputEl.click();
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error(error);
+    }
+  }, [inputEl]);
+
+  return { callback };
+};
+
+export const useUploadFiles: EnsembleActionHook<UploadFilesAction> = (
+  action?: UploadFilesAction,
+) => {
+  const [body, setBody] = useState<Record<string, unknown>>();
+  const [headers, setHeaders] = useState<Record<string, unknown>>();
+  const [status, setStatus] = useState<UploadStatus>("pending");
+  const [progress, setProgress] = useState<number>(0.0);
+
+  const onCompleteAction = useEnsembleAction(action?.onComplete);
+  const onErrorAction = useEnsembleAction(action?.onError);
+  const screenContext = useScreenContext();
+
+  useRegisterBindings(
+    {
+      body,
+      headers,
+      status,
+      progress,
+    },
+    action?.id,
+    {
+      setBody,
+      setHeaders,
+      setStatus,
+      setProgress,
+    },
+  );
+
+  const callback = useCallback(async (): Promise<void> => {
+    const apiModel = screenContext?.model?.apis?.find(
+      (model) => model?.name === action?.uploadApi,
+    );
+    if (!apiModel) return;
+
+    const progressCallback = (progressEvent: ProgressEvent): void => {
+      const percentCompleted =
+        (progressEvent.loaded * 100) / progressEvent.total;
+
+      setProgress(percentCompleted);
+    };
+
+    const files = evaluate(
+      screenContext as ScreenContextDefinition,
+      action?.files,
+    ) as FileList | undefined;
+    if (!files || files.length === 0) throw Error("Files not found");
+
+    const formData = new FormData();
+    if (files.length === 1)
+      formData.append(action?.fieldName ?? "files", head(files) as Blob);
+    else
+      for (let i = 0; i < files.length; i++) {
+        formData.append(action?.fieldName ?? `file${i}`, files[i]);
+      }
+
+    const apiModelBody = apiModel?.body ?? {};
+    for (const key in apiModelBody) {
+      const evaluatedValue = isExpression(apiModelBody[key])
+        ? evaluate(
+            screenContext as ScreenContextDefinition,
+            apiModelBody[key] as string,
+            action?.inputs,
+          )
+        : apiModelBody[key];
+      formData.append(key, evaluatedValue as string);
+    }
+
+    const apiUrl = evaluate(
+      screenContext as ScreenContextDefinition,
+      `\`${apiModel.uri}\``,
```

Comment:
Does work if the YAML string contains the backticks already?

Reply:
how can api uri already contain backticks? that would be syntax error
<img width="726" alt="image" src="https://github.com/EnsembleUI/ensemble-react/assets/32386713/10773d9f-02b2-46d0-86ec-b10dbfe12cd4">
or should we delete any existing backticks in the uri?